fuel-asm/src/macros.rs

```diff
@@ -713,6 +713,60 @@ macro_rules! op_unpack {
     () => {};
 }
 
+// Generate a method for checking that the reserved part of the
+// instruction is zero. This is private, as invalid instructions
+// cannot be constructed outside this crate.
+macro_rules! op_reserved_part {
```

Comment:
I think it is better to move this check on `unpack` level.

Something like this:
<img width="675" alt="image" src="https://github.com/FuelLabs/fuel-vm/assets/18346821/0ab4223c-1bd3-4e05-994a-9e7b89fa04c2">

Reply:
Can you elaborate why? It was my first instinct as well, but after I realized we can validate this at construction time and never construct invalid `Instruction`s.

Reply:
From a performance perspective. The check in the constructor repeats the logic in the `unpack`.

If we could remove the `unpack` method and store only required values like `RegId` or `Imm6`, we would be able to check it only once during creation and performance would be not affected=)

Reply:
The current approach works faster based on benchmarks than if the instruction is unpacked by default because(the memory layout of the unpacked variant is bigger).